### PR TITLE
Reimplement the DenoiserCov2D class using the BatchedRotCov2D Class

### PR DESCRIPTION
**Pull request by junchaoxia**
The PR includes the code for reimplementing DenoiserCov2D class using the BatchedRotCov2D Class so that there is no need to loading all 2D images into memory at once. The new implementation requires reading through the 2D images twice. The first time is for building the covariance matrix and the second one is for denoising noisy images and saving denoised ones.

### CODE REVIEW

File: src/aspire/source/__init__.py [hunk 1]
```diff
@@ -11,7 +10,6 @@
 from aspire.utils.filters import MultiplicativeFilter, PowerFilter
 from aspire.utils.coor_trans import grid_2d
 from aspire.source.xform import Multiply, Shift, Downsample, FilterXform, LinearIndexedXform, Pipeline, LinearPipeline
-from aspire.estimation.noise import WhiteNoiseEstimator
```
Comment (janden) [inline src/aspire/source/__init__.py:14]: Why is this (and the above removed import) in the current commit?

File: src/aspire/source/__init__.py [hunk 2]
```diff
@@ -172,6 +170,15 @@ def rots(self):
         """
         return self._rotations.as_dcm()
 
+    @property
```
Comment (janden) [inline src/aspire/source/__init__.py]: This is also not used in the current commit. Why is it added here?

File: src/aspire/source/relion.py [hunk 3]
```diff
@@ -11,47 +10,48 @@
...
 class RelionSource(ImageSource):
 
     @classmethod
-    def starfile2df(cls, filepath, data_folder=None, max_rows=None):
-        if data_folder is not None:
-            if not os.path.isabs(data_folder):
-                data_folder = os.path.join(os.path.dirname(filepath), data_folder)
-        else:
-            data_folder = os.path.dirname(filepath)
+    def starfile2df(cls, proj_folder, filepath, max_rows=None):
```
Comment (janden) [inline src/aspire/source/relion.py]: No space.

File: src/aspire/source/relion.py [hunk 4]
```diff
@@ -11,47 +10,48 @@
...
 
+        star_folder = proj_folder
+        if proj_folder is not None:
+            if not os.path.isabs(filepath):
```
Comment (janden) [inline src/aspire/source/relion.py]: So a relative STAR file path is assumed relative to project folder? Why?

File: src/aspire/source/relion.py [hunk 5]
```diff
@@ -11,47 +10,48 @@
...
         else:
             return df.iloc[:max_rows]
 
-    def __init__(self, filepath, data_folder=None, pixel_size=1, B=0, n_workers=-1, max_rows=None, memory=None):
+    def __init__(self, proj_folder, filepath, pixel_size=1, B=0, n_workers=-1, max_rows=None, memory=None):
         """
         Load STAR file at given filepath
-        :param filepath: Absolute or relative path to STAR file
-        :param data_folder: Path to folder w.r.t which all relative paths to .mrcs files are resolved.
+        :param proj_folder: Path to project folder w.r.t which all relative paths to all submodules..
```
Comment (janden) [inline src/aspire/source/relion.py]: Write out “with respect to.” What's a “submodule”? Remove extra period.

File: src/aspire/source/relion.py [hunk 6]
```diff
@@ -11,47 +10,48 @@
...
             If None, the folder corresponding to filepath is used.
+        :param filepath: Absolute or relative path to STAR file
```
Comment (janden) [inline src/aspire/source/relion.py]: Not relative to CWD but to project folder. Specify. Add period at the end.

File: src/aspire/source/relion.py [hunk 7]
```diff
@@ -11,47 +10,48 @@
...
-        df['__mrc_filepath'] = df['__mrc_filename'].apply(lambda filename: os.path.join(data_folder, filename))
+        df['__mrc_filepath'] = df['__mrc_filename'].apply(lambda filename: os.path.join(proj_folder, filename))
 
         if max_rows is None:
             return df
         else:
             return df.iloc[:max_rows]
 
-    def __init__(self, filepath, data_folder=None, pixel_size=1, B=0, n_workers=-1, max_rows=None, memory=None):
+    def __init__(self, proj_folder, filepath, pixel_size=1, B=0, n_workers=-1, max_rows=None, memory=None):
```
Comment (janden) [inline src/aspire/source/relion.py]: I don't quite see why the notion of a “project” should touch a `RelionSource`. What does it need to know other than where to find the MRC files?

File: src/aspire/commands/denoise.py [hunk 8]
```diff
@@ -67,10 +67,8 @@ def denoise(starfile_in, data_folder, starfile_out, pixel_size, max_rows, max_re
     var_noise = noise_estimator.estimate()
     # img_whitened = source.eval_filters(source.images())
     if denoise_method == 'CWF':
-        denoise_cov2d = DenoiserCov2D(source, basis, var_noise, ctf_info)
+        denoise_cov2d = DenoiserCov2D(source, basis, var_noise, ctf_info, starfile_out)
```
Comment (janden) [inline src/aspire/commands/denoise.py]: Why are we providing `starfile_out` to `DenoiserCov2D`? We should be able to denoise images *without* saving them to a STAR file.

File: src/aspire/denoising/denoiser_cov2d.py [hunk 9]
```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
...
     """
-    def __init__(self, src, basis, var_noise, ctf_info):
+    def __init__(self, src, basis, var_noise, ctf_info, star_file):
         """
         Initialize an object for denoising 2D images using Cov2D method
 
         :param src: The source object of 2D images with metadata
         :param basis: The basis method to expand 2D images
         :param var_noise: The estimated variance of noise
         :param ctf_info: Whether the CTF filters are included
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Isn't this available from the `src` object?

File: src/aspire/denoising/denoiser_cov2d.py [hunk 10]
```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
...
             raise NotImplementedError('Currently only fast FB method is supported')
-        self.coeffs_noise = None
-        self.coeffs_estim = None
-        self.ctf_idx = None
-        self.ctf_fb = None
         self.basis = basis
-        self.imgs_estim = None
+        self.ctf_info = ctf_info
+        self.star_file = star_file
+        self.starfile_path = os.path.dirname(star_file)
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Let's be consistent. It's either `star_file` or `starfile`.

File: src/aspire/denoising/denoiser_cov2d.py [hunk 11]
```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
...
-            uniq_ctf_filters = list(set(src.filters))
-            # Create the indices of CTF filters from all images
-            self.ctf_idx = np.array([uniq_ctf_filters.index(f) for f in src.filters])
-            # Evaluate CTFs in the FFB basis
-            self.ctf_fb = [f.fb_mat(basis) for f in uniq_ctf_filters]
         else:
             logger.info(f'CTF filters are not included in Cov2D denoising')
             # set all CTF filters to an identity filter
             self.ctf_idx = np.zeros(self.nimg, dtype=int)
             self.ctf_fb = [blk_diag_eye(blk_diag_partition(RadialCTFFilter().fb_mat(basis)))]
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Why isn't this handled in `BatchedCov2D`? It's not something that *only* arises here.

File: src/aspire/denoising/denoiser_cov2d.py [hunk 12]
```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
...
+        src = self.src
 
-        self.coeffs_noise = self.basis.evaluate_t(self.imgs_noise)
-
-        cov2d = RotCov2D(self.basis)
+        # Initialize the rotationally invariant covariance matrix of 2D images
+        # A fixed batch size is used to go through each image
+        cov2d = BatchedRotCov2D(self.src, self.basis, batch_size=512)
+        if not self.ctf_info:
+            cov2d.ctf_idx = self.ctf_idx
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Again, this shouldn't be needed. Better to handle it directly in `BatchedRotCov2D` (or even in `ImageSource`).

File: src/aspire/denoising/denoiser_cov2d.py [hunk 13]
```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
...
+        mean_coeffs_est = cov2d.get_mean()
+
+        covar_coeffs_est = cov2d.get_covar(noise_var=self.var_noise, mean_coeff=mean_coeffs_est,
+                                           covar_est_opt=covar_opt)
+        # Denoise the 2D images using the SPCAs from the rotationally invariant covariance matrix
+        # The batch size varies to fit the number of noisy images in each mrcs file
+        _, batch_size = src.group_unique_hist
+
+        img_start = 0
+        for ibatch in range(0, len(batch_size)):
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Why is this all done here? The images should be denoised *on demand* as part of a new `ImageSource`, not all at once.

File: src/aspire/denoising/denoiser_cov2d.py [hunk 14]
```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
...
+            # Convert Fourier-Bessel coefficients back into 2D images
+            logger.info(f'Converting Cov2D coefficients back to 2D images')
+            imgs_estim = self.basis.evaluate(coeffs_estim)
+            imgs_out = Image(imgs_estim)
+            mrcs_file = os.path.join(self.src.proj_folder, self.starfile_path, self.mrcs_out[img_start])
+            logger.info(f'Saving denoised images to {mrcs_file}')
+            imgs_out.save(mrcs_file, overwrite=overwrite)
+            img_start += batch_size[ibatch]
+
+    def _write_star_file(self):
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Again, we shouldn't have this here. Better to use the save functionality of the `ImageSource` class.

File: src/aspire/denoising/denoiser_cov2d.py [hunk 15]
```diff
@@ -18,72 +20,115 @@ class DenoiserCov2D(Denoiser):
...
         """
-        Return an ImageSource object with denoised images
+        Generate a name list of output mrcs files for denoised images
         """
-        src = ArrayImageSource(Image(self.imgs_estim))
-        return src
+        df = self.metadata
+        mrcs_fileout = []
+        for idx in range(len(df.index)):
+            mrcs_fileout.append(df['_rlnDenoisedImageName'][idx].split('/')[-1])
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Let's not use new fields unless it's absolutely necessary. This won't be interpretable by RELION. Better to store it in `_rlnImageName`.

**Review by janden — CHANGES_REQUESTED** [review]
Some things to discuss here.

File: src/aspire/source/__init__.py (hunk 1 above)
Comment (junchaoxia) [reply src/aspire/source/__init__.py:14]: I believe some code is moved to other place but the import statement was not. Just caught the redundancy and removed it.

File: src/aspire/source/__init__.py (hunk 2 above)
Comment (junchaoxia) [reply src/aspire/source/__init__.py]: Yes. This is for building batch size for another commit. The file is only committed  once.

File: src/aspire/source/relion.py (hunk 3 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: done.

File: src/aspire/source/relion.py (hunk 4 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: Yes. In the relion output, there is a globe project folder and under that each submodule such as particle picking and 2D classification creates a subfolder for each job to save the new star file and mrc files if necessary. In this way, new analysis can be started from different folders and will not mix up the input and output. I think it is a good way to follow in our package too and minimize the work when using the Relion results as inputs in our package.

File: src/aspire/source/relion.py (hunk 5 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: the `proj_folder` specifies the absolute path to the project folder with respect to all relative paths to the star files for different analysis such as denoising and 2D classification. Inside the star file, the paths for mrcs files will be relative to this  `proj_folder` too. In this way, no matter where is star file, the RelionSource function can find the mrc files using the path of `proj_folder`. This will be consistent with the Relion package.

File: src/aspire/source/relion.py (hunk 6 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: The filepath should include the relative path to project folder and input STAR file. I prefer to setting it as uniform format for all other modules as Cov3d too.

File: src/aspire/source/relion.py (hunk 7 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: As I replied for a comment above,  I tend to set a uniform format of all modules including cov2d, cov3d and others. The `proj_folder` specifies the absolute path to the project folder with respect to all relative paths to the star files for different analysis such as denoising and 2D classification. Inside the star file, the paths for mrcs files will be relative to this `proj_folder` too. In this way, no matter where is star file, the RelionSource function can find the mrc files using the path of `proj_folder` and the entry in the star file. This will be more consistent with the Relion package. The `filepath` should include the relative path to project folder and input STAR file.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 13 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: I tried to use the batch size according to the number of noisy images in each mrcs file. In the way, we can use multi-process mode easily in future by assigning one batch (one set of images in the same mrcs file) to one process and no need to worry about the order of file writing. I probably will change the function in the `ImageSource`  after the discussion.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 14 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Yes. I might change or move to `ImageSource` class.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 15 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: It should be fine to use the same name if we just replace the original noise image.

File: src/aspire/commands/denoise.py (hunk 8 above)
Comment (junchaoxia) [reply src/aspire/commands/denoise.py]: This star file is an output for future analysis of other modules such as 2D classification. I think it might be useful to keep the similar format with Relion.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 9 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: The src object only has the star file for input mrc files. This star file is for the output of denoised images.

File: src/aspire/source/relion.py (hunk 4 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: I reverted back to the original format using the data folder and created an new issue #141 for future discussion.

File: src/aspire/source/relion.py (hunk 5 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: Same as above. I reverted back to the original format using the data folder and created an new issue #141 for future discussion.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 11 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Moved to BatchedCov2D in new commit f83b805.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 13 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: A `DenoisedImageSource` class was added for this purpose in the new commit 23ca25c.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 14 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Same as above, a  `DenoisedImageSource` class was added for this purpose in the new commit 23ca25c.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 15 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Changed to use the same name `_rlnImageName` in the new star file.

File: src/aspire/source/relion.py (hunk 6 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: Same as above. I reverted back to the original format using the data folder and created an new issue #141 for future discussion.

File: src/aspire/source/relion.py (hunk 7 above)
Comment (junchaoxia) [reply src/aspire/source/relion.py]: Same as above. I reverted back to the original format using the data folder and created an new issue #141 for future discussion.

File: src/aspire/commands/denoise.py (hunk 8 above)
Comment (junchaoxia) [reply src/aspire/commands/denoise.py]: Removed from the DenoiseCov2Dand used a new  `DenoisedImageSource` class for this in the new commit 23ca25c.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 9 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Same as above, the star file was removed from the DenoiseCov2D and used a new  `DenoisedImageSource` class for this in the new commit 23ca25c.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 10 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Same as above, the star file was removed from the DenoiseCov2D and used a new  `DenoisedImageSource` class for this in the new commit 23ca25c.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 12 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: As above, it is moved to `BatchedCov2D` in new commit f83b805.

File: azure-pipelines.yml [hunk 16]
```diff
@@ -1,4 +1,26 @@
 jobs:
+- job:
```
Comment (janden) [inline azure-pipelines.yml:2]: These changes need to be in a different PR.

File: src/aspire/commands/denoise.py [hunk 17]
```diff
@@ -65,12 +63,12 @@ def denoise(starfile_in, data_folder, starfile_out, pixel_size, max_rows, max_re
     logger.info(f'Whiten the noise of images from the noise estimator')
     source.whiten(noise_estimator.filter)
     var_noise = noise_estimator.estimate()
-    # img_whitened = source.eval_filters(source.images())
+
+    # source.create_star(starfile_out, batch_size=512)
```
Comment (janden) [inline src/aspire/commands/denoise.py]: Remove commented-out code.

File: src/aspire/denoising/__init__.py [hunk 18]
```diff
@@ -21,26 +20,12 @@ def __init__(self, src):
 
     def denoise(self):
         """
-        Denoise 2D images
+        Precompute for Denoiser and DenoisedImageSource for 2D images
         """
         raise NotImplementedError('subclasses must implement this')
 
-    def to_src(self):
+    def images(self):
```
Comment (janden) [inline src/aspire/denoising/__init__.py]: Signature (arguments) should match the one in the subclass.

File: src/aspire/denoising/denoised_src.py [hunk 19]
```diff
@@ -0,0 +1,90 @@
...
+        """
+        Initialize a denoised ImageSource object from original ImageSource of noisy images
+
+        :param src: Original ImageSource object storing noisy images
+        :param denoiser: A Denoiser object for specifying a method for denoising
+        """
+
+        super().__init__(src.L, src. n, dtype=src.dtype,metadata=src._metadata.copy())
+        self._im = None
+        # self._metadata=src._metadata.copy()
```
Comment (janden) [inline src/aspire/denoising/denoised_src.py]: Remove commented-out code.

File: src/aspire/denoising/denoised_src.py [hunk 20]
```diff
@@ -0,0 +1,90 @@
...
+
+    def __init__(self, src, denoiser):
+        """
+        Initialize a denoised ImageSource object from original ImageSource of noisy images
+
+        :param src: Original ImageSource object storing noisy images
+        :param denoiser: A Denoiser object for specifying a method for denoising
+        """
+
+        super().__init__(src.L, src. n, dtype=src.dtype,metadata=src._metadata.copy())
```
Comment (janden) [inline src/aspire/denoising/denoised_src.py]: Space after comma.

File: src/aspire/denoising/denoised_src.py [hunk 21]
```diff
@@ -0,0 +1,90 @@
...
+        im = np.empty((self.L, self.L, len(indices)))
+
+        logger.info(f'Loading {len(indices)} images complete')
+        for istart in range(start, end, batch_size):
+            imgs_denoised = self.denoiser.images(istart, batch_size)
+            im = imgs_denoised.data
+
+        return Image(im)
+
+    def _create_star(self, starfile_filepath, batch_size=512):
```
Comment (janden) [inline src/aspire/denoising/denoised_src.py]: Remove unused method.

File: src/aspire/denoising/denoiser_cov2d.py [hunk 22]
```diff
@@ -18,72 +15,73 @@ class DenoiserCov2D(Denoiser):
...
         default_opt = {'shrinker': 'frobenius_norm', 'verbose': 0, 'max_iter': 250,
             'iter_callback': [], 'store_iterates': False, 'rel_tolerance': 1e-12,
             'precision': 'float64'}
         covar_opt = fill_struct(covar_opt, default_opt)
-        mean_coeffs_est = cov2d.get_mean(self.coeffs_noise, self.ctf_fb, self.ctf_idx)
-
-        covar_coeffs_est = cov2d.get_covar(self.coeffs_noise, self.ctf_fb, self.ctf_idx, mean_coeffs_est,
-                                               noise_var=self.var_noise, covar_est_opt=covar_opt)
+        # Calculate the mean and covariance for the rotationally invariant covariance matrix of 2D images
+        self.mean__est = self.cov2d.get_mean()
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Shouldn't this be `mean_est`? Does this not cause the code to crash?

File: src/aspire/denoising/denoiser_cov2d.py [hunk 23]
```diff
@@ -18,72 +15,73 @@ class DenoiserCov2D(Denoiser):
...
+        :param istart: the index of starting image
+        :param batch_size: The batch size for processing images
+        :return: an `Image` object with denoised images
         """
-        src = ArrayImageSource(Image(self.imgs_estim))
-        return src
+        src = self.src
+
+        # Denoise one batch size of 2D images using the SPCAs from the rotationally invariant covariance matrix
+
```
Comment (janden) [inline src/aspire/denoising/denoiser_cov2d.py]: Remove blank line.

File: src/aspire/io/starfile.py [hunk 24]
```diff
@@ -183,22 +185,46 @@ def save_star(image_source, starfile_filepath, batch_size=1024, overwrite=False)
...
-            logger.info(f'Saving ImageSource[{i_start}-{i_end}] to {mrcs_filepath}')
-            im = image_source.images(start=i_start, num=num)
-            im.save(mrcs_filepath, overwrite=overwrite)
-
-            df['_rlnImageName'][i_start: i_end] = pd.Series(['{0:06}@{1}'.format(j + 1, mrcs_filepath) for j in range(num)])
-
+        else:
+            # save all images into multiple mrc files in batch size
+            for i_start in np.arange(0, image_source.n, batch_size):
+
```
Comment (janden) [inline src/aspire/io/starfile.py]: Remove blank line.

File: src/aspire/io/starfile.py [hunk 25]
```diff
@@ -183,22 +185,46 @@ def save_star(image_source, starfile_filepath, batch_size=1024, overwrite=False)
...
+                i_end = min(image_source.n, i_start + batch_size)
+                num = i_end - i_start
+                mrcs_filename = os.path.splitext(os.path.basename(starfile_filepath)
```
Comment (janden) [inline src/aspire/io/starfile.py]: Don't split the line just before parenthesis.

File: tests/test_simulation.py [hunk 26]
```diff
@@ -60,14 +60,22 @@ def testSimulationEigen(self):
...
-                [-1.15402077e-06, -2.49849709e-05, -3.51658906e-04, -2.21575261e-03, -7.83315487e-03, -9.44795180e-03, -4.07636259e-03, -9.02186439e-04],
-                [-1.88737249e-05, -1.91418396e-04, -1.09021540e-03, -1.02020288e-03,  1.39411855e-02,  8.58035963e-03, -5.54619730e-03, -3.86377703e-03],
-                [-1.21280536e-04, -9.51461843e-04, -3.22565017e-03, -1.05731178e-03,  2.61375736e-02,  3.11595201e-02,  6.40814053e-03, -2.31698658e-02],
-                [-2.44067283e-04, -1.40560151e-03, -6.73082832e-05,  1.44160679e-02,  2.99893934e-02,  5.92632964e-02,  7.75623545e-02,  3.06570008e-02],
-                [-1.53507499e-04, -7.21709803e-04,  8.54929152e-04, -1.27235036e-02, -5.34382043e-03,  2.18879692e-02,  6.22706190e-02,  4.51998860e-02],
-                [-3.00595184e-05, -1.43038429e-04, -2.15870258e-03, -9.99002904e-02, -7.79077187e-02, -1.53395887e-02,  1.88777559e-02,  1.68759506e-02],
-                [ 3.22692649e-05,  4.07977635e-03,  1.63959339e-02, -8.68835449e-02, -7.86240026e-02, -1.75694861e-02,  3.24984640e-03,  1.95389288e-03]
+                [-1.67666201e-07, -7.95741380e-06, -1.49160041e-04, -1.10151654e-03,
```
Comment (janden) [inline tests/test_simulation.py]: Why do these have to be changed?

**Review by janden — APPROVED** [review]
Looks good. Just a few minor things.

File: azure-pipelines.yml (hunk 16 above)
Comment (junchaoxia) [reply azure-pipelines.yml:2]: The related commit has be cherry-picked to the develop branch .

File: src/aspire/commands/denoise.py (hunk 17 above)
Comment (junchaoxia) [reply src/aspire/commands/denoise.py]: Done.

File: src/aspire/denoising/__init__.py (hunk 18 above)
Comment (junchaoxia) [reply src/aspire/denoising/__init__.py]: Done.

File: src/aspire/denoising/denoised_src.py (hunk 19 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoised_src.py]: Done.

File: src/aspire/denoising/denoised_src.py (hunk 20 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoised_src.py]: Done.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 22 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Good catch. It will recalculate it if the value is None.

File: src/aspire/denoising/denoiser_cov2d.py (hunk 23 above)
Comment (junchaoxia) [reply src/aspire/denoising/denoiser_cov2d.py]: Done.

File: src/aspire/io/starfile.py (hunk 24 above)
Comment (junchaoxia) [reply src/aspire/io/starfile.py]: Done.

File: src/aspire/io/starfile.py (hunk 25 above)
Comment (junchaoxia) [reply src/aspire/io/starfile.py]: Done.

File: tests/test_simulation.py (hunk 26 above)
Comment (junchaoxia) [reply tests/test_simulation.py]: This is to make the representation of covariance more accurate. The related commit has been reverted and make it be addressed in more details in future issue.
